### PR TITLE
Sliding Hash Join Algorithm (SWHJ)

### PR DESCRIPTION
**Pull request by metesynnada**
### **Rationale for this change**

This PR aims to introduce a new incremental processing algorithm, the Sliding Window Hash Join (SWHJ), into our query processing engine. The SWHJ algorithm is designed for joining two possibly infinite streams while preserving the order of one of the input streams. By incrementally building/maintaining its hash table, SWHJ offers significant performance and functionality improvements over the traditional hash join algorithm. It provides a solution that is both stream-friendly and order-preserving, effectively handling scenarios where a sliding window constraint is applicable.

### **What changes are included in this PR?**

**Implementation of the Sliding Window Hash Join (SWHJ)**: This algorithm incrementally builds its hash table, making it suitable for handling infinite streams while preserving input ordering. The implementation includes logic to manage the sliding window constraint in the join operation, ensuring efficient processing of incoming rows from both streams without needing to materialize the entire input.

### **Are these changes tested?**

Yes

### **Are there any user-facing changes?**

These changes introduce the Sliding Window Hash Join (SWHJ) algorithm as a new method for join operations. Users now have the ability to use this algorithm when a sliding window constraint is applicable to their use case. The algorithm's ability to handle infinite streams while preserving the input order provides significant performance benefits over the traditional hash join operation.

### CODE REVIEW

File: datafusion/core/src/physical_plan/joins/sliding_hash_join.rs [hunk 1]
```diff
@@ -0,0 +1,1480 @@
...
+        join_type: &JoinType,
+        null_equals_null: bool,
+        left_sort_exprs: Vec<PhysicalSortExpr>,
+        right_sort_exprs: Vec<PhysicalSortExpr>,
+        mode: StreamJoinPartitionMode,
+    ) -> Result<Self> {
+        let left_schema = left.schema();
+        let right_schema = right.schema();
+
+        // Error out if no "on" contraints are given:
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/joins/sliding_hash_join.rs:180]: constraints

File: datafusion/core/src/physical_plan/joins/sliding_hash_join.rs [hunk 2]
```diff
@@ -0,0 +1,1480 @@
...
+        // Build the join schema from the left and right schemas:
+        let (schema, column_indices) =
+            build_join_schema(&left_schema, &right_schema, join_type);
+
+        // Initialize the random state for the join operation:
+        let random_state = RandomState::with_seeds(0, 0, 0, 0);
+
+        let filter_state = Arc::new(Mutex::new(IntervalCalculatorInnerState::default()));
+
+        let output_ordering = calculate_sliding_join_output_order(
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/joins/sliding_hash_join.rs:199]: we can use `calculate_join_output_ordering` here I guess.

File: datafusion/core/src/physical_plan/joins/sliding_hash_join.rs [hunk 3]
```diff
@@ -0,0 +1,1480 @@
...
+            ),
+            Some(
+                self.right_sort_exprs
+                    .to_vec()
+                    .iter()
+                    .map(|o| PhysicalSortRequirement::from(o.clone()))
+                    .collect::<Vec<_>>(),
+            ),
+        ]
+    }
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/joins/sliding_hash_join.rs:310]: we can use below code instead
```rust
        vec![
            Some(PhysicalSortRequirement::from_sort_exprs(
                &self.left_sort_exprs,
            )),
            Some(PhysicalSortRequirement::from_sort_exprs(
                &self.right_sort_exprs,
            )),
        ]
```

File: datafusion/core/src/physical_plan/joins/sliding_hash_join.rs [hunk 4]
```diff
@@ -0,0 +1,1480 @@
...
+        combine_join_equivalence_properties(
+            self.join_type,
+            self.left.equivalence_properties(),
+            self.right.equivalence_properties(),
+            left_columns_len,
+            &self.on,
+            self.schema(),
+        )
+    }
+
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/joins/sliding_hash_join.rs:327]: We can implement ordering_equivalence_properties for this executor also, using combine_join_ordering_equivalence_properties util.

File: datafusion/core/src/physical_plan/joins/sliding_hash_join.rs [hunk 5]
```diff
@@ -0,0 +1,1480 @@
...
+                &self.right,
+                &self.filter,
+            )? {
+            (left_sorted_filter_expr, right_sorted_filter_expr, graph)
+        } else {
+            return Err(DataFusionError::Internal("AsymmetricHashJoin can not operate without both sides are not prunning tables.".to_owned()));
+        };
+
+        let on_left = self.on.iter().map(|on| on.0.clone()).collect::<Vec<_>>();
+        let on_right = self.on.iter().map(|on| on.1.clone()).collect::<Vec<_>>();
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/joins/sliding_hash_join.rs:385]: You can use 
```rust
        let (on_left, on_right) = self.on.iter().cloned().unzip();
```
here

File: datafusion/core/src/physical_plan/joins/sliding_hash_join.rs [hunk 6]
```diff
@@ -0,0 +1,1480 @@
...
+                JoinStreamState::PullProbe => {
+                    loop {
+                        match ready!(self.right_stream.poll_next_unpin(cx)) {
+                            Some(Ok(batch)) => {
+                                // Update metrics for polled batch
+                                self.metrics.right.input_batches.add(1);
+                                self.metrics.right.input_rows.add(batch.num_rows());
+
+                                // Check if batch suits interval calculation criteria
+                                let continue_pooling =
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/joins/sliding_hash_join.rs:684]: continue_polling

File: datafusion/core/src/physical_plan/joins/sliding_hash_join.rs [hunk 7]
```diff
@@ -0,0 +1,1480 @@
...
+                                    !is_batch_suitable_interval_calculation(
+                                        &self.right_sorted_filter_expr,
+                                        &batch,
+                                    )?;
+                                self.probe_buffer.candidate_buffer.push(batch);
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/joins/sliding_hash_join.rs:689]: You can move this line to above `let continue_pooling` assignment

File: datafusion/core/src/physical_plan/joins/sliding_hash_join.rs [hunk 8]
```diff
@@ -0,0 +1,1480 @@
...
+                            Some(Ok(batch)) => {
+                                if batch.num_rows() == 0 {
+                                    continue;
+                                }
+                                self.metrics.left.input_batches.add(1);
+                                self.metrics.left.input_rows.add(batch.num_rows());
+                                let array_ref = build_order
+                                    .expr
+                                    .evaluate(&batch)?
+                                    .into_array(batch.num_rows());
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/joins/sliding_hash_join.rs:742]: You can use last row in the batch for the evaluation

File: datafusion/core/src/physical_plan/joins/sliding_hash_join.rs [hunk 9]
```diff
@@ -0,0 +1,1568 @@
...
+use datafusion_execution::memory_pool::MemoryConsumer;
+use datafusion_execution::TaskContext;
+use datafusion_physical_expr::intervals::ExprIntervalGraph;
+use datafusion_physical_expr::{OrderingEquivalenceProperties, PhysicalSortRequirement};
+
+use ahash::RandomState;
+use futures::{ready, Stream, StreamExt};
+use parking_lot::Mutex;
+
+/// Sliding hash join implementation for sliding window joins.
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/joins/sliding_hash_join.rs:70]: Let's write a longer docstring with more explanations, like the docstring of `SlidingNestedLoopJoinExec`.

File: datafusion/core/src/physical_plan/joins/sliding_hash_join.rs [hunk 10]
```diff
@@ -0,0 +1,1568 @@
...
+                            self.join_type,
+                            &self.column_indices,
+                        )?;
+
+                        // Prune the internal state of the build hash joiner:
+                        build_hash_joiner.prune_internal_state(prune_length)?;
+                        maybe_batch
+                    } else {
+                        None
+                    };
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/joins/sliding_hash_join.rs:918]: The corresponding block in the SNLJ code doesn't have the `prune_length > 0` guard. Do we really need it here, in contrast to the SNLJ case?

**Review by ozankabak — COMMENTED** [review]
I left two minor, easy-to-fix comments. Except those, my main concern is about the duplication of state machine code with SNLJ. It is almost the same code: Except calling specialized local functions (e.g. `join_with_probe_batch`) and one line about managing offsets, the code seems the same. Let's discuss how we can apply DRY here.

As a side note (not to be included in this PR), some of the SHJ docstrings are now old; e.g. `build_side_determined_results`, `join_with_probe_batch`. They also have crate level visibility which is not needed anymore. Let's create a "maintenance" PR to fix these things and send upstream.

**Review by ozankabak — APPROVED** [review]

**Pull request merged**